src/uct/ib/dc/dc_mlx5.c

```diff
@@ -483,36 +549,343 @@ static void uct_dc_mlx5_iface_cleanup_dcis(uct_dc_mlx5_iface_t *iface)
+#endif
+}
+
+ucs_status_t uct_ib_mlx5_set_srq_dc_params(uct_dc_mlx5_iface_t *iface)
+{
+#if HAVE_DECL_MLX5DV_CONTEXT_FLAGS_DEVX
+    uint32_t in[UCT_IB_MLX5DV_ST_SZ_DW(set_xrq_dc_params_entry_in)] = {};
+    uint32_t out[UCT_IB_MLX5DV_ST_SZ_DW(set_xrq_dc_params_entry_out)] = {};
+    int ret;
+
+    UCT_IB_MLX5DV_SET(set_xrq_dc_params_entry_in, in, pkey_table_index, iface->super.super.super.pkey_index);
+    UCT_IB_MLX5DV_SET(set_xrq_dc_params_entry_in, in, mtu, iface->super.super.config.path_mtu);
+    UCT_IB_MLX5DV_SET(set_xrq_dc_params_entry_in, in, sl, iface->super.super.super.config.sl);
+    UCT_IB_MLX5DV_SET(set_xrq_dc_params_entry_in, in, reverse_sl, iface->super.super.super.config.sl);
+    UCT_IB_MLX5DV_SET(set_xrq_dc_params_entry_in, in, cnak_reverse_sl, iface->super.super.super.config.sl);
+    UCT_IB_MLX5DV_SET(set_xrq_dc_params_entry_in, in, ack_timeout, iface->super.super.config.timeout);
+    UCT_IB_MLX5DV_SET64(set_xrq_dc_params_entry_in, in, dc_access_key, UCT_IB_KEY);
+    UCT_IB_MLX5DV_SET(set_xrq_dc_params_entry_in, in, xrqn, iface->rx.srq_num);
+    UCT_IB_MLX5DV_SET(set_xrq_dc_params_entry_in, in, opcode,
+                      UCT_IB_MLX5_CMD_OP_SET_XRQ_DC_PARAMS_ENTRY);
+
+    ret = mlx5dv_devx_obj_modify(iface->rx.srq, in, sizeof(in), out, sizeof(out));
+    if (ret) {
+        return UCS_ERR_IO_ERROR;
+    }
+#endif
+    return UCS_OK;
+}
+
 static ucs_status_t
 uct_dc_mlx5_init_rx(uct_rc_iface_t *rc_iface,
                     const uct_rc_iface_config_t *rc_config)
 {
     uct_dc_mlx5_iface_t *iface = ucs_derived_of(rc_iface, uct_dc_mlx5_iface_t);
 
-#if IBV_EXP_HW_TM_DC
     uct_dc_mlx5_iface_config_t *config = ucs_derived_of(rc_config,
                                                         uct_dc_mlx5_iface_config_t);
     if (UCT_RC_MLX5_TM_ENABLED(&iface->super)) {
-         struct ibv_exp_create_srq_attr srq_attr      = {};
-         struct ibv_exp_srq_dc_offload_params dc_op   = {};
-
-         iface->super.super.progress = uct_dc_mlx5_iface_progress_tm;
-
-         dc_op.timeout    = rc_iface->config.timeout;
-         dc_op.path_mtu   = rc_iface->config.path_mtu;
-         dc_op.pkey_index = rc_iface->super.pkey_index;
-         dc_op.sl         = rc_iface->super.config.sl;
-         dc_op.dct_key    = UCT_IB_KEY;
+        struct ibv_exp_create_srq_attr srq_attr = {};
+        ucs_status_t status;
+#ifdef HAVE_STRUCT_IBV_EXP_CREATE_SRQ_ATTR_DC_OFFLOAD_PARAMS
+        struct ibv_exp_srq_dc_offload_params dc_op = {};
+
+        dc_op.timeout    = rc_iface->config.timeout;
+        dc_op.path_mtu   = rc_iface->config.path_mtu;
+        dc_op.pkey_index = rc_iface->super.pkey_index;
+        dc_op.sl         = rc_iface->super.config.sl;
+        dc_op.dct_key    = UCT_IB_KEY;
+
+        srq_attr.comp_mask         = IBV_EXP_CREATE_SRQ_DC_OFFLOAD_PARAMS;
+        srq_attr.dc_offload_params = &dc_op;
+#endif
+        status = uct_dc_mlx5_init_rx_tm(iface, &config->super, &srq_attr);
```

Comment:
seems uct_ib_mlx5_set_srq_dc_params() should be called from inside uct_dc_mlx5_init_rx_tm(), since "dc_mlx5_rx_tm" also includes setting DC parameters. And also move the EXP code to uct_dc_mlx5_init_rx_tm()..